File: game-of-life/kotlin-arrow-game-of-life/src/main/kotlin/GameOfLife.kt
import Console.puts
import GamePlay.calculateNextGeneration
import arrow.fx.IO
import arrow.fx.IO.Companion.just
import arrow.fx.extensions.fx

object GameOfLife {
    fun gameOfLife(): IO<Matrix> = welcome()
            .flatMap { readInitialMatrix() }
            .flatMap { gameLoop(it) }

    private fun welcome(): IO<Unit> = puts("Welcome to the Conway's Game of Life!")

    private fun readInitialMatrix(): IO<Matrix> = just(
            listOf(
                    listOf(Dead, Alive, Dead),
                    listOf(Dead, Alive, Dead),
                    listOf(Dead, Alive, Dead)
            )
    )

    private fun print(matrix: Matrix): IO<Matrix> = puts(showMatrix.run { matrix.show() })
            .flatMap { just(matrix) }

    private fun gameLoop(matrix: Matrix): IO<Matrix> = IO.fx {
        val currentMatrix = !print(matrix)
        val nextMatrix = !calculateNextGeneration(currentMatrix)
        !gameLoop(nextMatrix)
    }
}

File: game-of-life/typescript-fat-arrow-ts-game-of-life/src/GameOfLife.ts
import {Matrix} from "./Data";
import {print, readInitialMatrix, welcome} from "./GameInteraction";
import {calculateNextGeneration} from "./GamePlay";

export const gameOfLife: Promise<Matrix> =
    welcome
        .then(() => readInitialMatrix
            .then((matrix) => print(matrix)
                .then((matrix) => gameLoop(matrix))
            )
        )

const gameLoop: (matrix: Matrix) => Promise<Matrix> = (matrix: Matrix) =>
    calculateNextGeneration(matrix)
        .then((matrix) => print(matrix)
            .then(matrix => gameLoop(matrix))
        )
